src/TemplateScope.ts

```diff
@@ -3,50 +3,96 @@
 // ----------------------------------------------------------------------------
 
 import * as assert from 'assert';
-import { Utilities } from '../extension.bundle';
+import { Json, Utilities } from '../extension.bundle';
+import { CachedValue } from './CachedValue';
 import { templateKeys } from './constants';
 import { IParameterDefinition } from "./IParameterDefinition";
+import { IResource } from './IResource';
 import * as TLE from "./TLE";
 import { UserFunctionDefinition } from './UserFunctionDefinition';
 import { UserFunctionNamespaceDefinition } from "./UserFunctionNamespaceDefinition";
 import { IVariableDefinition } from './VariableDefinition';
 
-export enum ScopeContext {
+export enum TemplateScopeKind {
     TopLevel = "TopLevel",
     ParameterDefaultValue = "ParameterDefaultValue",
-    UserFunction = "UserFunction"
+    UserFunction = "UserFunction",
+    NestedDeploymentWithInnerScope = "NestedDeploymentWithInnerScope",
+    NestedDeploymentWithOuterScope = "NestedDeploymentWithOuterScope",
 }
 
 /**
  * Represents the scoped access of parameters/variables/functions at a particular point in the template tree.
  */
-export class TemplateScope {
-    /**
-     * Constructor
-     */
+export abstract class TemplateScope {
+    private _parameterDefinitions: CachedValue<IParameterDefinition[] | undefined> = new CachedValue<IParameterDefinition[] | undefined>();
+    private _variableDefinitions: CachedValue<IVariableDefinition[] | undefined> = new CachedValue<IVariableDefinition[] | undefined>();
+    private _functionDefinitions: CachedValue<UserFunctionNamespaceDefinition[] | undefined> = new CachedValue<UserFunctionNamespaceDefinition[] | undefined>();
```

Comment:
>UserFunctionNamespaceDefinition [](start = 46, length = 31)

Nit: I don't fully know the codebase, but it looks a bit weird for UserFunctionNameSpaceDefinition to be accessed via class rather than interface as all other xxxDefinitions, no?

Reply:
Considered changing that (does look weird), but decided there wasn't really any advantage to doing it. For Typescript, it's a lot easer to switch to an interface when it's actually needed, so would rather keep things simple until the extra abstraction's actually needed.